### ISSUE TITLE
Add np.argval() to complement np.argsort()?

### ISSUE THREAD

**Issue by mspacek** [numpy/numpy#9052]
I often need to find where the values in array `v` can be found in the reference array `a`. `np.searchsorted(a, v)` is very useful for exactly this reason, but only if `a` is sorted. If `a` is not sorted, there's a code pattern @joferkington posted [here](http://stackoverflow.com/a/8251668) that does the trick without resorting to something slowish like `np.array([ int(np.where(a == val)[0]) for val in v ])`. For my own use, I've named this pattern `argval`:
```python
def argval(a, v):
    """Find indices into input array `a` where values in array `v` match those in `a`.
       a : input array
       v : array of values to find in a

       This should return exactly the same as:

       `np.array([ int(np.where(a == val)[0]) for val in v ])`

       but faster. Adapted from http://stackoverflow.com/a/8251668"""
    a, v = np.asarray(a), np.asarray(v)
    if not set(v).issubset(a):
        raise ValueError("values array %r is not a subset of input array %r" % (v, a ))
    asortis = a.argsort()
    return asortis[a[asortis].searchsorted(v)]

```
I'm currently only using this for arrays of integers, but it works with strings too. Floats might cause problems in the subset test.

If I submit this as a pull request, might it be considered, or am I reinventing something that already exists in numpy?

**Comment by eric-wieser** [comment]
Am I right in thinking that this is more of an `index`, where `out[i] == v.index(a[i])`?

I'm not a fan of the subset approach - in particular, it might be handy to return `-1` or something for `out[i]` where `a[i] not in v`

**Comment by eric-wieser** [comment]
So I think I would be +1 on this feature in some form being a good addition to `arraysetops`, with a better name and some discussion over how to handle missing values

**Comment by mspacek** [comment]
> Am I right in thinking that this is more of an index, where out[i] == v.index(a[i])?

Hm, I don't quite understand what you mean. `out` should have the same length as `v`. Wouldn't it be more like `out = a.index(v)`? So maybe the functional form could be called `np.index(a, v)`?

Also, I'm not sure it belongs in `arraysetops`, because it preserves order and allows for repetition of values. Don't the setops generally sort everything?

**Comment by eric-wieser** [comment]
You're right, I do mean `a.index(v[i])`. `np.index` seems far too generic a name, especially given we already have `np.ndindex` which means something totally different. `np.index_of` is a little better.

And yes, perhaps `arraysetops` is not so good a fit.

**Comment by mspacek** [comment]
How about `np.find`, or `np.argmatch`?

As for returning a placeholder like `-1` for `v[i] not in a` instead of testing for `v.issubset(a)`, perhaps that could be controlled via a kwarg? The problem with `-1` is that it's actually a valid index, so unless you test for it in `out` yourself, it might do unexpected things when you use `out` to index into something like `a`.

**Comment by mspacek** [comment]
I just noticed that `np.searchsorted` now has a `sorter` kwarg, so this also works:
```python
asortis = np.argsort(a)
return asortis[a.searchsorted(v, sorter=asortis)]
```

**Comment by shoyer** [comment]
Here is a variation on the same problem that has been upvoted several hundred times on StackOverflow: [Is there a Numpy function to return the first index of something in an array?](http://stackoverflow.com/questions/432112/is-there-a-numpy-function-to-return-the-first-index-of-something-in-an-array)

I'd like to consider if we can come up with a good solution for both problems. Possibly `np.find_all` (which might use `searchsorted`, or `where`, depending on the size of the second argument, and always return an array of indices) and `np.find_first` (which would shortcut after finding one match, and return a single index or raise an error).

**Comment by seberg** [comment]
You could possibly have a `np.equal.find_first(..., order="C")` ufunc method (or possibly find N first), which would work for any `...->?` signature ufunc (actually, could work for any ufunc, since you can always add a cast to bool at the end I guess. The python workaround for huge arrays would probably by `nditer` based chunking on the python side (which is mostly the same thing as you would do in C anyway).
